src/query/query_executor.c

```diff
@@ -7678,8 +7689,38 @@ qexec_intprt_fnc (THREAD_ENTRY * thread_p, XASL_NODE * xasl, XASL_STATE * xasl_s
 
   while ((xb_scan = qexec_next_scan_block_iterations (thread_p, xasl)) == S_SUCCESS)
     {
+      if (xasl->max_iterations != -1)
+	{
+	  assert (xasl->curr_spec->type == TARGET_LIST);
+	  assert (xasl->curr_spec->s_id.type == S_LIST_SCAN);
+
+	  cte_start_new_iteration = true;
+	  recursive_iterations = 1;
+	}
+
       while ((ls_scan = scan_next_scan (thread_p, &xasl->curr_spec->s_id)) == S_SUCCESS)
 	{
+	  if (xasl->max_iterations != -1)
+	    {
+	      if (cte_start_new_iteration)
+		{
+		  recursive_iterations++;
+		  if (recursive_iterations >= xasl->max_iterations)
+		    {
+		      max_recursive_iterations_reached = true;
+		      break;
+		    }
+
+		  curr_iteration_last_cursor = CTE_CURR_ITERATION_LAST_TUPLE (xasl);
```

Comment:
Both recursive and non_recursive XASL are not available at this stack level, so such assert is not possible here.
They are available only at the calling function (qexec_execute_cte), where the code flow is straight: first loop-> sets common_list_optimization = true; (after the lists are concatenated)
second loop (and the rest) : recursive_part->max_iterations is set (otherwise remains -1). I don't think such assert is required in qexec_execute_cte.